scipy/special/special/config.h

```diff
@@ -2,83 +2,99 @@
+#define M_LOG2E 1.44269504088896340736
 #endif
 
-#if __has_include(<cmath>)
-#include <cmath>
-#else
+#ifndef M_LOG10E
+#define M_LOG10E 0.434294481903251827651
+#endif
 
-namespace std {
+#ifndef M_LN2
+#define M_LN2 0.693147180559945309417
+#endif
 
-SPECFUN_HOST_DEVICE inline double abs(double num) { return thrust::abs(num); }
+#ifndef M_LN10
+#define M_LN10 2.30258509299404568402
+#endif
 
-SPECFUN_HOST_DEVICE inline double exp(double num) { return thrust::exp(num); }
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
 
-SPECFUN_HOST_DEVICE inline double fma(double x, double y, double z) { return x * y + z; }
+#ifndef M_PI_2
+#define M_PI_2 1.57079632679489661923
+#endif
 
-SPECFUN_HOST_DEVICE inline double log(double num) { return thrust::log(num); }
+#ifndef M_PI_4
+#define M_PI_4 0.785398163397448309616
+#endif
 
-SPECFUN_HOST_DEVICE inline double sqrt(double num) { return thrust::sqrt(num); }
+#ifndef M_1_PI
+#define M_1_PI 0.318309886183790671538
+#endif
 
-SPECFUN_HOST_DEVICE inline bool isnan(double num) { return thrust::isnan(num); }
+#ifndef M_2_PI
+#define M_2_PI 0.636619772367581343076
+#endif
 
-} // namespace std
+#ifndef M_2_SQRTPI
+#define M_2_SQRTPI 1.12837916709551257390
+#endif
 
+#ifndef M_SQRT2
+#define M_SQRT2 1.41421356237309504880
 #endif
 
-#if __has_include(<limits>)
+#ifndef M_SQRT1_2
+#define M_SQRT1_2 0.707106781186547524401
+#endif
 
-#include <limits>
+#include <cuda/std/cmath>
+#include <cuda/std/limits>
 
-#else
+// Fallback to global namespace for functions unsupported on NVRTC Jit
+#ifdef _LIBCUDACXX_COMPILER_NVRTC
+#include <cuda_runtime.h>
+#endif
 
 namespace std {
 
-template <typename T>
-class numeric_limits;
+SPECFUN_HOST_DEVICE inline double abs(double num) { return cuda::std::abs(num); }
 
-template <>
-class numeric_limits<double> {
-  public:
-    SPECFUN_HOST_DEVICE static constexpr double infinity() { return CUDART_INF; }
+SPECFUN_HOST_DEVICE inline double exp(double num) { return cuda::std::exp(num); }
 
-    SPECFUN_HOST_DEVICE static constexpr double quiet_NaN() { return CUDART_NAN; }
-};
+SPECFUN_HOST_DEVICE inline double log(double num) { return cuda::std::log(num); }
 
-} // namespace std
+SPECFUN_HOST_DEVICE inline double sqrt(double num) { return cuda::std::sqrt(num); }
 
-#endif
+SPECFUN_HOST_DEVICE inline bool isnan(double num) { return cuda::std::isnan(num); }
 
-#if __has_include(<complex>)
+SPECFUN_HOST_DEVICE inline bool isfinite(double num) { return cuda::std::isfinite(num); }
 
-#include <complex>
+SPECFUN_HOST_DEVICE inline double pow(double x, double y) { return cuda::std::pow(x, y); }
 
+SPECFUN_HOST_DEVICE inline double sin(double x) { return cuda::std::sin(x); }
+
+// Fallback to global namespace for functions unsupported on NVRTC
+#ifndef _LIBCUDACXX_COMPILER_NVRTC
+SPECFUN_HOST_DEVICE inline double floor(double x) { return cuda::std::floor(x); }
+SPECFUN_HOST_DEVICE inline double fma(double x, double y, double z) { return cuda::std::fma(x, y, z); }
```

Comment:
`cuda::std::` is all supported with NVRTC. The trick is you just need to add the cuda/std headers to the include path for NVRTC.

Reply:
Thanks, very cool. I'm still wrapping my head around things. My impression was based on [cuda/std/detail/libcxx/cmath#L395](https://github.com/NVIDIA/cccl/blob/ef2d09b101bcdbabc994bcd526be58541c3e1242/libcudacxx/include/cuda/std/detail/libcxx/include/cmath#L395)

```C++
...
#ifndef _LIBCUDACXX_COMPILER_NVRTC


using ::fpclassify;
using ::isnormal;
using ::isgreater;
using ::isgreaterequal;
using ::isless;
using ::islessequal;
using ::islessgreater;

using ::isunordered;

using ::float_t;
using ::double_t;

using ::fabs;
using ::fabsf;
using ::floor;
using ::floorf;
...
```

where these functions are only pulled from the global namespace into `cuda::std` if this isn't running with NVRTC. The things in this block are the ones I haven't been able to directly use with CuPy. I haven't studied the code carefully enough to see how these functions are still provided on NVRTC as long as the include paths are set up correctly.

I suppose the include paths need to be changed on the CuPy side. We're just getting started, but I'm sure things will fall into place as we get rolling. It would be awesome if we could just conditionally alias std to cupy::std and have things just work in both SciPy and CuPy.